Profile Overrides (RFC #2282 Part 1)

ehuss opened the pull request:
Profile Overrides (RFC #2282 Part 1)

WIP: Putting this up before I dig into writing tests, but should be mostly complete.  I also have a variety of questions below.

This implements the ability to override profiles for dependencies and build scripts.  This includes a general rework of how profiles work internally. Closes #5298.

Profile overrides are available with `profile-overrides` set in `cargo-features` in the manifest.

Part 2 is to implement profiles in config files (to be in a separate PR).

General overview of changes:

- `Profiles` moved to `core/profiles.rs`. All profile selection is centralized there.
- Removed Profile flags `test`, `doc`, `run_custom_build`, and `check`.
- Removed `Profile` from `Unit` and replaced it with two enums: `CompileMode` and `ProfileFor`.  This is the minimum information needed to compute profiles at a later stage.
- Also removed `rustc_args`/`rustdoc_args` from `Profile` and place them in `Context`.  This is currently not very elegant because it is a special case, but it works. An alternate solution I considered was to leave them in the `Profile` and add a special uber-override layer.  Let me know if you think it should change.
- Did some general cleanup in `generate_targets`.

## Misc Fixes
- `cargo check` now honors the `--release` flag.  Fixes #5218.
- `cargo build --test` will set `panic` correctly for dependences. Fixes #5369.
- `cargo check --tests` will no longer include bins twice (once as a normal check, once as a `--test` check).  It only does `--test` check now.
    - Similarly, `cargo check --test name` no longer implicitly checks bins.
- Examples are no longer considered a "test".  (See #5397). Consequences:
    - `cargo test` will continue to build examples as a regular build (no change).
    - `cargo test --tests` will no longer build examples at all.
    - `cargo test --all-targets` will no longer build examples as tests, but instead build them as a regular build (now matches `cargo test` behavior).
    - `cargo check --all-targets` will no longer check examples twice (once as
      normal, once as `--test`).  It now only checks it once as a normal
      target.

## Questions
- Thumbs up/down on the general approach?
- The method to detect if a package is a member of a workspace should probably be redone.  I'm uncertain of the best approach.  Maybe `Workspace.members` could be a set?
- `Hash` and `PartialEq` are implemented manually for `Profile` only to avoid matching on the `name` field.  The `name` field is only there for debug purposes. Is it worth it to keep `name`?  Maybe useful for future use (like #4140)?
- I'm unhappy with the `Finished` line summary that displays `[unoptimized + debuginfo]`.  It doesn't actually show what was compiled.  Currently it just picks the base "dev" or "release" profile.  I'm not sure what a good solution is (to be accurate it would need to potentially display a list of different options).  Is it ok?  (See also #4140 for the wrong profile name being printed.)
- Build-dependencies use different profiles based on whether or not `--release` flag is given.  This means that if you want build-dependencies to always use a specific set of settings, you have to specify both `[profile.dev.build_override]` and `[profile.release.build_override]`.  Is that reasonable (for now)?  I've noticed some issues (like #1774, #2234, #2424) discussing having more control over how build-dependencies are handled.
- `build --bench xxx` or `--benches` builds dependencies with dev profile, which may be surprising.  `--release` does the correct thing.  Perhaps print a warning when using `cargo build` that builds benchmark deps in dev mode?
- Should it warn/error if you have an override for a package that does not exist?
- Should it warn/error if you attempt to set `panic` on the `test` or `bench` profile?

## TODO
- I have a long list of tests to add.
- Address a few "TODO" comments left behind.

matklad commented on src/cargo/core/compiler/context/unit_dependencies.rs:
```diff
@@ -38,27 +39,33 @@ fn deps_of<'a, 'b, 'cfg>(
+        let unit_deps = compute_deps(unit, cx, deps, profile_for)?;
+        let to_insert: Vec<_> = unit_deps.iter().map(|&(unit, _)| unit).collect();
+        deps.insert(*unit, to_insert);
+        for (unit, profile_for) in unit_deps {
```
If I add `assert_eq!(unit.profile_for, profile_for);` it is not triggered, so, perhaps, we don't need to return pairs after all?

matklad commented:
>Thumbs up/down on the general approach?

:+1: :+1: :+1: 

>The method to detect if a package is a member of a workspace should probably be redone.

I think current implementation is OK: linear scan is perfectly fine until you have hundreds of members.

>Is it worth it to keep name?

I have no opinion here :) 

> I'm unhappy with the Finished line summary

It still works correctly when no overrides are present, and otherwise it is a minor detail, so I suggest to punt on this for know: current solution seems at least OK to me!

> Build-dependencies use different profiles based on whether or not --release flag is given. 

Yeah, that seems generally wrong, but OK for now, because it more or less matches the current behavior. I would say that this is probably an instance where we do need a separate, not dev/not release profile, but let's deal with that later.

> build --bench xxx or --benches builds dependencies with dev profile,

Hm, is this the current behavior? It definitely is a bug!

> Should it warn/error if you have an override for a package that does not exist?

Let's give at least a warning if it's not too hard: typoes are likely in the names of packages, especially with `-` vs `_` distinction. 

> Should it warn/error if you attempt to set panic on the test or bench profile?

It is allowed now, right? Then I think error is probably not feasible, but the warning seems useful!

matklad commented on src/cargo/core/profiles.rs:
```diff
@@ -0,0 +1,394 @@
+}
+
+impl ProfileFor {
+    pub fn all_values() -> Vec<ProfileFor> {
```
Mincooptimization, but here and in `CompileMode::all_modes`, we could return `&'static [Stuff]`, to save some allocations.

matklad commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -88,13 +92,19 @@ pub struct Context<'a, 'cfg: 'a> {
+    /// command-line for `cargo rustc` and `cargo rustdoc`.  These commands
+    /// only support one target, but we don't want the args passed to any
+    /// dependencies.
+    pub extra_compiler_args: HashMap<Unit<'a>, Vec<String>>,
```
Perhaps we can use `Option<(Unit, Vec<String>)>` instead? Let's also document that this is overloaded in the sense that this can be arguments for different tools, `rustc` or `rustdoc`, depending on the command.

matklad commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -534,307 +548,308 @@ impl CompileFilter {
+    // Helper for creating a Unit struct.
+    let new_unit =
+        |pkg: &'a Package, target: &'a Target, mode: CompileMode, profile_for: ProfileFor| {
+            let actual_profile_for = if profile_for != ProfileFor::Any {
```
Just a personal taste, but I would name this just `profile_for` to shadow the input parameter, so that, down the line, we can't use the wrong `profile_for`.

ehuss commented:
Thanks @matklad for doing such a detailed review so quickly!

Unfortunately, I've discovered a pretty fundamental flaw in how it works.  Units are not being properly de-duplicated.  Example: test→lib and bin→lib, the lib gets compiled twice (once as a regular dependency, and again as a test dependency).  But if you don't have `panic` set they get compiled with the exact same settings!

I think Profile really must be a part of Unit.  I had a previous attempt that had it part of the Unit, but I was concerned about the huge amount of copying.  However, I was using Strings before (instead of InternedString), and it seemed much worse at the time (require lots of `clone()` calls).  I'm going to rewrite it to use this approach.  `Profile` is only 24 bytes on my system, so it shouldn't be too bad.  On the plus side, it should simplify some things.

Sorry for the churn, I'll try to fix this today and address all your comments.

ehuss commented:
I've pushed the update that now embeds the profile inside `Unit`.  It works quite a bit better now.  I fear there are still some hidden gotchas, so I'll need to spend a fair amount of time making tests.  I left a long comment in there about a subtle issue with `cargo test`, that I think I'm going to revisit because it's not quite right.  Getting proper `panic` support is turning out to be very tricky.  Doc tests are also turning out to be very tricky, because it greedily links against any libs it finds in `Context.libraries`.

bors commented:
:umbrella: The latest upstream changes (presumably #5392) made this pull request unmergeable. Please resolve the merge conflicts.

matklad commented on src/cargo/core/compiler/context/unit_dependencies.rs:
```diff
@@ -29,8 +29,13 @@ pub fn build_unit_dependencies<'a, 'cfg>(
+        // cleared, and avoid building the lib thrice (once with `panic`, once
+        // without, once for --test).  In particular, the lib included for
+        // doctests and examples are `Build` mode here.
+        let profile_for = if unit.mode.is_any_test() || cx.build_config.test {
```
Hm, I am genuinely not sure what the correct behavior would be here? Suppose we are compiling the integration test, which only executes the binary in the same package. It is reasonable to assume that the binary might be compiled with `panic=abort`, but the test itself should be compiled with `panic=unwind`, so common dependencies should be compiled twice. 

Perhaps the best path forward is to preserve the current, buggy, behavior, land this PR, and then fix the behavior in a follow-up PR?

matklad commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -568,15 +569,34 @@ fn generate_targets<'a>(
+                // `panic` set.
+                //
+                // As a consequence, Examples and Binaries get compiled
+                // without `panic` set.  This probably isn't a bad deal.
```
Ah yeah, this is exactly what I was thinking about... Hm, compiling binaries with different panic setting actually seems like a big deal to me...

ehuss replied:
Yea, it's a tricky issue, I was going to ask about it.  The immediate problem is that doc tests attempt to link with everything in `Context.libraries`.  I'm uncertain what the easiest way to fix that would be.  One thought I had would be to change `run_doc_tests` to only grab the dependencies for the `CompileMode::Doctest` unit and remove `Context.libraries`.  There would be some other minor things to accommodate that, but I think it should work.  I can do it now or later, let me know.  It would remove this weird special-case.

matklad replied:
Would it be difficult to preserve exactly the behavior (and bugs :) ) we have today? I think this would be the best way forward for now, because changes which both add unstable features and fix/modify stable behavior are tricky...

ehuss replied:
Yea, it should be preserving the old behavior.  I have a few changes I haven't pushed, yet, that ensure that.  I'm keeping a list of behavior changes, and I'll post it soon.  A few are clearly bug fixes, but a couple are questionable.  We can review them when I post it.

ehuss commented:
The CI failures are due to the introduction of rust-lang/rust#49289 (starting with nightly 2018-04-15).  `cargo check` will now emit `rmeta` files for all targets.  This will actually address #3624.  However, I'm uncertain how the test should be structured now.  It's trying to verify the "freshness", but the behavior has changed between rust versions.  How do you typically accommodate different rustc versions?

ehuss commented:
(I've updated the first comment with some other Misc Fixes that probably aren't too controversial, but please review them.  Sorry for the long comment, I just wanted to be thorough.)

The latest push includes some tests (`profile_targets.rs`) that illustrate what I was asking about before in regards to a way to verify exactly which commands get run, but ignoring the order.  I don't think it would be possible to achieve the same level of verification with the existing functions (in particular, verifying what doesn't get run).  Let me know if you would prefer it to be done in a different way, though.  One alternative thought I had would be to have an option for `with_json` to ignore the order of messages, and add some more information to `ArtifactProfile` to be able to distinguish which profile is used.

## Questions
- The `doc` profile does not seem to be used.  Internally it is assigned to running `rustdoc`, but `rustdoc` does not use any flags from the profile.  Is that intended?  Perhaps we could deprecate it (or maybe just remove it from the docs)?  There are about 24 crates on crates.io that set it for whatever reason.
- I just want to double-check... I've been using the term "profile override" in a few places, but I'm starting to feel that might not be clear (compared to `[profile.dev]` is overriding the default dev profile).  "package profile override" would be more accurate, but is maybe a little wordy?

## Behavior Changes to Discuss
The following are new changes in behavior that I'm uncertain if we should keep or change.

- `cargo build --release --all-targets` now compiles libs and bins twice (once for benchmarks, once for unit-tests).  Previously the de-dupe logic would only compile them once (since they are essentially identical).  This is because with `CompileMode` in `Unit` it distinguishes between a Bench and Test, so it doesn't get de-duped.  I could maybe add a special case for this?

- If a `build.rs` is needed for different profiles, it will now be run separately for each one.  For example, `build --all-targets` will run `build.rs` once for the `dev` targets, and once for the `bench` targets (because the `opt-level` is different between them).  I think this is more-correct behavior, but may be a surprising change to some.  Thoughts?

## Panic Changes and Issues
Getting `panic` correct has been quite tricky.  This is the list of changes and issues I'm currently aware of.  These assume you have `panic` set in some profile.

- `cargo build` with tests or benches was previously failing (```error: the crate `foo` is compiled with the panic strategy `abort` which is incompatible with this crate's strategy of `unwind` ```).  This is now fixed.

- `cargo test` builds binaries without `panic` set.  This is a tricky issue.  If this is fixed, then the `lib` would get built three times (once without panic for tests, once with panic for bins, once with `--test` as a unit-test).  The bigger problem is that `rustdoc` for doc-tests would fail to run because it would pick up both copies of the lib.  See comment previously about needing to change `Context.libraries` to fix this.  Leave as-is?  Fix now/later?

- `build.rs` scripts and their dependencies have `panic` cleared because the panic-clearing logic thinks it is a compiler plugin (see `build_base_args`).  However, I don't think `build.rs` scripts should be treated the same as compiler plugins/proc-macros.  Would it be worth it to fix this?  The change is fairly easy -- I would probably add `for_plugin` (`for_compiler`?) to `Target` and adjust `walk_used_in_plugin_map` to use that instead of `for_host`.  This is a minor issue, but I was a little surprised to see it in one of my test cases.
The consequences are:
    - `build.rs` scripts are always built without panic.
    - If you have a dependency that is shared as a regular dependency and a build dependency, and you run `cargo test` or `cargo bench`, the dependency will be built twice, but with the same settings.  Internally it thinks one is "dev without panic" (for tests) and the other is regular "dev" (for `build.rs`).  It works fine, but slows things down since the shared dependency is built twice.

## Known Bugs
- `DEBUG` env var in `build.rs` is incorrect if profile `debug=0`.  #5370.  Will fix in a separate PR.
- Having a global `build_config.release` flag causes some issues.  I'm uncertain how difficult this is to fix.  Generally getting rid of the global `build_config.release` would be the correct approach.  Thoughts?  Address now or later?  I think this only affects `cargo build`, and the workaround is to pass `--release` to it.  Consequences:
    - `PROFILE` env var in `build.rs` is incorrect for `cargo build --bench name`, it says it is `"debug"` when in reality it is `"release"`.
    - `cargo build --bench xxx` will build dependencies in `dev` profile, which may be surprising (#4929).
- `cargo check` only ever uses `dev` or `release` profile.  For example, `cargo check --test foo` uses `dev` instead of `test`.  I'm uncertain how any of the profile flags might affect `rustc` when running `check`.

matklad commented:
>The latest push includes some tests (profile_targets.rs) that illustrate what I was asking about before in regards to a way to verify exactly which commands get run, but ignoring the order. I don't think it would be possible to achieve the same level of verification with the existing functions (in particular, verifying what doesn't get run)

Yeah, that seems reasonable. Let's add `with_stderr_unordered` perhaps which compares *sets* of lines?

matklad commented:
>However, I'm uncertain how the test should be structured now. It's trying to verify the "freshness", but the behavior has changed between rust versions. How do you typically accommodate different rustc versions?

I think we can add 
```
// If you are reading this, and 1.27.0 is already stable, feel free to just remove this if!
if !is_nightly() { return; } 
```

matklad commented:
>The doc profile does not seem to be used. Internally it is assigned to running rustdoc, but rustdoc does not use any flags from the profile. Is that intended? Perhaps we could deprecate it (or maybe just remove it from the docs)? There are about 24 crates on crates.io that set it for whatever reason.

Yep, I believe we can safely deprecate this profile. That is, leave the syntax in Cargo.toml, give a warning about it, and remove traces of doc profile from the code. Eventually I think we'll do the same for `test` and `bench` profiles, but they *are* used, so we'll deal with that separately. 

> "package profile override" would be more accurate, but is maybe a little wordy?

I personally think about it as just "package profile". We need `override` in the surface syntax just to namespace this feature. I wonder if a sweeter syntax is possible? `[profiles.dev.for.image]` perhaps?

>cargo build --release --all-targets now compiles libs and bins twice 

I think it is important to remove this duplication. Could we get rid of the `Bench` compile mode? 


>If a build.rs is needed for different profiles, it will now be run separately for each one. For example, build --all-targets will run build.rs once for the dev targets, and once for the bench targets (because the opt-level is different between them). I think this is more-correct behavior, but may be a surprising change to some. Thoughts?

Hm, so currently build_scripts are compiled only with `dev` or `release` profiles. I suggest we keep this behavior, for three reasons:

* No change means that we won't break someone's code,
* If we remove all profiles except `dev` and `test`, we'll effectively have to change it back,
* Test fiddle with `panic` setting, and this I think should not affect build scripts. 

Long term, I believe the current behavior of using both `dev` and `release` for build scripts is also wrong. I think we need a separate use-visible profile for build scripts and their dependencies, so that `cargo build && cargo build --release` does not compile build scripts twice. 

I don't think there should be *any* sharing between the build script and the main code. Especially for cross-compilation, it seems unreasonable to use the same profile for both.

ehuss commented:
> Yep, I believe we can safely deprecate this profile.

Would you like that now, or in a separate PR (since this one is getting a little long)?

> [profiles.dev.for.image] perhaps?

Hm, that is a little clearer.  Not sure if anyone else wants to weigh in?

> I think it is important to remove this duplication. Could we get rid of the Bench compile mode?

It would be nice, because once the `Unit` is constructed, it is not needed.  However, there would need to be a way for `generate_targets` and `generate_default_targets` to know that we are `cargo bench` to select the correct targets and profiles.  I could pass a flag in through `CompileOptions`, what do you think?  An alternate solution is to just swap the mode in `generate_targets` when the Unit is made which would be a one-line change to fix this.

> Hm, so currently build_scripts are compiled only with dev or release profiles. I suggest we keep this behavior, for three reasons:

This is how it works (before and after this change) for *compiling* the `build.rs` script.  I was talking about executing it.  In the new code, it is run twice because the `OPT_LEVEL` is different for different targets' dependencies.  This goes back to, should `cargo build --bench xxx` build the dependencies of a benchmark in `dev` mode?  I guess I left it sort this in a halfway state.  I see two options.  One is to restore the old behavior (benches get `dev` for custom-built dependencies), and I can fix that by just checking the global `release` flag for *running* `build.rs`, which I think is an easy change.  The other is to actually fix that problem, which might be more difficult (involves removing the use of the global `release` flag).  I'm fine with restoring the old behavior, this is an edge case (building benchmarks with `cargo build` without `--release` flag).  I'm also fine with trying to fix it properly.  Or other options (like not even trying to build benches with `cargo build` if you don't set `--release`?).

> I think we need a separate use-visible profile for build scripts and their dependencies

Isn't that what `build-override` is essentially?  I think I mentioned it earlier, that this pseudo-profile already exists, but you have to define it twice (`profile.dev.build-override` and `profile.release.build-override`), and it has a weird relationship with other overrides.

This brings up a issue that was nagging at me.  I didn't fully understand the reasoning for the override precedence in the RFC.  Since other rules take precedence over `build-override`, it seems like that can yank away the ability to independently control how a build script is compiled vs how the dependency itself is compiled.   There is a comment in the RFC thread that you and Manishearth had separately about this, but it didn't make sense to me.  The build-dependencies are always kept independent of the final binary, even if we removed the precedence.  Changing the precedence could give the user more control.  (Or even adding something like `[profile.scripts.overrides.somedep]` which would be very easy to do.)  I guess my rambling question is, why not give independent control?

> Could you elaborate why do you think so?

The only reason this exception is made is to ensure the `panic` flag is cleared (because plugins are linked with things that can't have it set).  `build.rs` scripts don't have that limitation since they aren't linked with the compiler.  I'm fine with clearing it for `build.rs`.  I'm uncertain how likely it is to be a problem for someone.  Considering it has always been this way, and things built with `panic=abort` is very rare to start with, it might not bother anyone.  However, if `panic` is set, it could cause shared dependencies to be needlessly built twice.

> I think it's fine to leave this duplication as is, but, if we are to fix it, then perhaps " Internally it thinks one is "dev without panic" (for tests) and the other is regular "dev" (for build.rs)." is the bit that needs adjustments?

Correct, it's a small, trivial change to say "panic = None" for build scripts in the profile builder.

matklad commented:
>Would you like that now, or in a separate PR (since this one is getting a little long)?

A separate one. I'd prefer to merge this sooner, and not to fix  **all** profile bugs at once :) 

>Hm, that is a little clearer. Not sure if anyone else wants to weigh in?

I personally don't care about syntax *at this stage*. There's already `override` in there, so let's not change it for now. 

>I could pass a flag in through CompileOptions, what do you think? An alternate solution is to just swap the mode in generate_targets when the Unit is made which would be a one-line change to fix this.

Another option is to split `CompileMode` into two enums: one to be used with `CompileOptions`, which  corresponds to which cli options were passed, and one to be used with `Units`. But "swapping the mode" solution also sounds fine with me!

>I was talking about executing it

Oh, right! I've completely misunderstood your original comment then! The new behavior indeed seems more correct to me as well!

>Isn't that what build-override is essentially? I think I mentioned it earlier, that this pseudo-profile already exists, but you have to define it twice (profile.dev.build-override and profile.release.build-override), and it has a weird relationship with other overrides.

Yep, it's almost like build-override, with two important differences: you don't need to override it twice, and, by default, build profile is the same for `cargo build` and `cargo build --release`. 

>The build-dependencies are always kept independent of the final binary, even if we removed the precedence.

Hm, I think we though that dependencies are shared between the library and the build script. Because they are independent, priority for build_overrides indeed does not make sense.

ehuss commented:
> Hm, I think we though that dependencies are shared between the library and the build script. Because they are independent, priority for build_overrides indeed does not make sense.

OK, let me know if you want to iterate on this now or later.  It's pretty easy to change.

Let me know if there's anything else.  I think I've gone through the issues I can think of.

matklad commented on src/cargo/core/compiler/context/unit_dependencies.rs:
```diff
@@ -38,27 +50,33 @@ fn deps_of<'a, 'b, 'cfg>(
     unit: &Unit<'a>,
     cx: &Context<'a, 'cfg>,
     deps: &'b mut HashMap<Unit<'a>, Vec<Unit<'a>>>,
+    profile_for: ProfileFor,
```
I feel uneasy about this `profile_for` argument. This function is memoized, so the sequence of calls `deps_of(u, cx, deps, ProfileFor::Any); deps_of(u, cx, deps, ProfileFor::TestDependency)` will return the same result each time (and the result might be different if we swap the order of calls).

ehuss replied:
I understand what you mean.  In practice, that's never a problem for `TestDependency` since it only exists for clearing `panic`, and you'll never ask for a unit *with* panic set as a `TestDependency`.  

I'm trying to think of a scenario where it would matter for `CustomBuild`, but I can't think of any.  It might matter if we later add a "build-script"-specific profile, and you have an override deep in a dependency chain.

Adding it to the map would require a separate de-duplication phase, which I think would add ~15 lines of code.

Let me know if you think it's important.  I'd lean towards not changing it unless there's a testable case.  Or to make it easier to add a build-script profile later if there's a strong chance of that.

matklad replied:
I think I'm fine with it: I would love to refactor code in such a way that this problem just not arises, or at least to put some sort of run-time assertion, but I have no ideas of how to actually do this. Storing pairs in the `HashMap` would be worse than the status quo. 

Let's perhaps just add a small comment here though?

matklad commented on src/cargo/core/compiler/context/unit_dependencies.rs:
```diff
@@ -273,28 +283,65 @@ fn maybe_lib<'a, 'cfg>(unit: &Unit<'a>, cx: &Context<'a, 'cfg>) -> Option<Unit<'
 /// of work is still returned. `None` is only returned if the package has no
 /// build script.
-fn dep_build_script<'a, 'cfg>(unit: &Unit<'a>, cx: &Context<'a, 'cfg>) -> Option<Unit<'a>> {
+fn dep_build_script<'a>(cx: &Context, unit: &Unit<'a>) -> Option<(Unit<'a>, ProfileFor)> {
```
Some function in this module now use cx, unit arg order, and some use (unit, cx). Let's unify them to one order? I don't have any preference for a specific order, as long as it is an order :-)

ehuss replied:
Yea, I derped.  I had an earlier version that I had removed Context, but when I added it back I should have kept it the same.

matklad commented on tests/testsuite/profile_targets.rs:
```diff
@@ -0,0 +1,703 @@
+foo custom build PROFILE=release DEBUG=false OPT_LEVEL=3
+[RUNNING] `rustc --crate-name foo src[/]lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C panic=abort -C codegen-units=2 [..]
+[RUNNING] `rustc --crate-name foo src[/]main.rs --crate-type bin --emit=dep-info,link -C opt-level=3 -C panic=abort -C codegen-units=2 [..]
+[FINISHED] release [optimized] [..]
```
I guess I don't understand something about panic handling... How this even works? We link `bar`, which is compiled with `panic=unwind` into `foo`, which is compiled with `panic=abort` and `rustc` does not complain. Is it the expected behavior? Testing locally, I see that `unwind -> abort` direction works, and `abort -> unwind` fails. Does this mean that it is always safe to compile dependencies with `panic = unwind`? 


This probably has nothing to do with the PR though, because the behavior has not changed here.

ehuss replied:
Hm, yea, I don't know much how panic handling is implemented.  I did a test where I attempt to panic in a dependency that was compiled with unwind, and it still aborts.  It looks like it just affects whether `libpanic_abort` or `libpanic_unwind` is linked.  I'm guessing since the API is the same, the dependencies don't know the difference.

matklad replied:
>It looks like it just affects whether libpanic_abort or libpanic_unwind is linked. I

I would guess that it also affects generation of landing pads? So we might be getting a bunch of dead code here?

@alexcrichton could you explain how mixing panic modes is supposed to work?

alexcrichton replied:
Ah yes the observed behavior here is expected, we , for example, ship panic=unwind libstd to everyone but still allow compiling with panic=abort

alexcrichton commented on src/cargo/core/profiles.rs:
```diff
@@ -0,0 +1,395 @@
+    }
+
+    fn validate_packages(&self, shell: &mut Shell, packages: &HashSet<&str>) -> CargoResult<()> {
+        if let Some(ref toml) = self.toml {
```
Stylistically I'm always personally a fan of reducing indent here:

```rust
let toml = match self.toml {
    Some(ref toml) => toml,
    None => return Ok(()),
};
// ...
```

alexcrichton commented on src/cargo/core/profiles.rs:
```diff
@@ -0,0 +1,395 @@
+        if let Some(ref toml) = self.toml {
+            if let Some(ref overrides) = toml.overrides {
+                for key in overrides.keys().filter(|k| k.as_str() != "*") {
+                    if !packages.contains(key.as_str()) {
```
I think this is something we'll want to note on the tracking issue, I'd expect the `packages` set here to be more like a `HashSet<PackageId>` (or a `Resolve`) and the keys of `overrides` here would be a `PackageIdSpec` in theory

alexcrichton replied:
Basically yeah. We've got the idea of a "package id specification" which is intended to describe a particular crate in a crate graph unbambiguously, and this would allow you to, for example, customize the compilation of `foo` from crates.io but not customize `foo` from github.com/example/foo (or something like that)

bors commented:
:umbrella: The latest upstream changes (presumably #5415) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.4524), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/372262361?utm_source=github_status&utm_medium=notification)
Approved by: matklad
Pushing c24a09772c2c1cb315970dbc721f2a42d4515f21 to master...
